tools/i2ctransfer.8
.TH i2ctransfer 8 "May 2022"
.SH "NAME"
i2ctransfer \- send user-defined I2C messages in one transfer

.SH SYNOPSIS
.B i2ctransfer
.RB [ -a ]
.RB [ -b ]
.RB [ -f ]
.RB [ -v ]
.RB [ -y ]
.I i2cbus desc
.RI [ data ]
.RI [ desc
.RI [ data ]]
.RI ...
.br
.B i2ctransfer
.B -h
.br
.B i2ctransfer
.B -V

.SH DESCRIPTION
.B i2ctransfer
is a program to create I2C messages and send them combined as one transfer.
For read messages, the contents of the received buffers are printed to stdout, one line per read message.
.br
Please note the difference between a
.I transfer
and a
.I message
here.
A transfer may consist of multiple messages and is started with a START condition and ends with a STOP condition as described in the I2C specification.
Messages within the transfer are concatenated using the REPEATED START condition which is described there as well.
There are some advantages of having multiple messages in one transfer.
First, some devices keep their internal states for REPEATED START but reset them after a STOP.
Second, you cannot get interrupted during one transfer, but it might happen between multiple transfers.
Interruption could happen on hardware level by another I2C master on the bus, or at software level by another I2C user who got its transfer scheduled between yours.
This program helps you to create proper transfers for your needs.

.SH OPTIONS
.TP
.B -a
Allow using addresses between 0x00 - 0x07 and 0x78 - 0x7f. Not recommended.
.TP
.B -b
Print data from read messages as binary data. Disables
.I -v
(verbose). Useful for piping into a pretty printing tool like
.B hexdump
or for redirecting into a file to be analyzed later.
.TP
.B -f
Force access to the device even if it is already busy.
By default,
.B i2ctransfer
will refuse to access a device which is already under the control of a kernel driver.
Using this flag is dangerous, it can seriously confuse the kernel driver in question.
It can also cause
.B i2ctransfer
to silently write to the wrong register.
So use at your own risk and only if you know what you're doing.
.TP
.B -h
Display the help and exit.
.TP
.B -v
Enable verbose output.
It will print infos about all messages sent, i.e. not only for read messages but also for write messages.
.TP
.B -V
Display the version and exit.
.TP
.B -y
Disable interactive mode.
By default,
.B i2ctransfer
will wait for a confirmation from the user before messing with the I2C bus.
When this flag is used, it will perform the operation directly.
This is mainly meant to be used in scripts.

.SH ARGUMENTS
.PP
The first parameter
.I i2cbus
indicates the number or name of the I2C bus to be used.
This number should correspond to one of the busses listed by
.B i2cdetect -l.

.PP
The next parameter is one or multiple
.I desc
blocks.
The number of blocks is limited by the Linux Kernel and defined by I2C_RDWR_IOCTL_MAX_MSGS (42 as of v4.10).
.I desc
blocks are composed like this:

.I [inpst]{r|w}<length_of_message>[@address]

.TP
.B [inpst]
If supported, specifies optional MESSAGE MODIFIER FLAGS. See the section below for details.
.TP
.B {r|w}
specifies if the message is read or write
.TP
.B <length_of_message>
specifies the number of bytes read or written in this message.
It is parsed as an unsigned 16 bit integer, but note that the Linux Kernel applies an additional upper limit (8192 as of v4.10).
For read messages to targets which support SMBus Block transactions, it can also be '?', then the target will determine the length.
.TP
.B [@address]
specifies the 7-bit address of the chip to be accessed for this message, and is an integer.
If omitted, reuse the previous address.
Normally, addresses outside the range of 0x08-0x77 and addresses with a kernel driver attached to them will be blocked.
This can be overridden with
.I -a
(all) or
.I -f
(force).
Be very careful when using these!
10-bit addresses are currently not supported at all.

.PP
If the I2C message is a write, then a
.I data
block with the data to be written follows.
It consists of
.I <length_of_message>
bytes which can be marked with the usual prefixes for hexadecimal, octal, etc.
To make it easier to create larger data blocks easily, the data byte can have a suffix.

.TP
=
keep value constant until end of message (i.e. 0= means 0, 0, 0, ...)
.TP
+
increase value by 1 until end of message (i.e. 0+ means 0, 1, 2, ...)
.TP
-
decrease value by 1 until end of message (i.e. 0xff- means 0xff, 0xfe, 0xfd, ...)
.TP
p
use value as seed for an 8 bit pseudo random sequence (i.e. 0p means 0x00, 0x50, 0xb0, ...)

.SH "MESSAGE MODIFIER FLAGS"
.PP
These optional flags can be used to change the handling of a message when it is sent on the bus.
They allow you to deviate from the I2C standard to work around device issues.
They should be used with care, as they can confuse your I2C bus.
Note that not all flags (or any) may be supported by your particular hardware and your kernel version.

.TP
.B i
ignore NACK from client, treat them as ACK.
.TP
.B n
in a read message, master ACK/NACK bit is skipped.
.TP
.B p
emit a STOP after the message.
.TP
.B s
skip repeated start.
.TP
.B t
toggle read/write bit.

.SH EXAMPLES
.PP
On bus 0, from an EEPROM at address 0x50, read 8 byte from offset 0x64
(first message writes one byte to set the memory pointer to 0x64, second message reads from the same chip):
.nf
.RS
# i2ctransfer 0 w1@0x50 0x64 r8
.RE
.fi
.PP
For the same EEPROM, at offset 0x42 write 0xff 0xfe ... 0xf0
(one write message; first byte sets the memory pointer to 0x42, 0xff is the first data byte, all following data bytes are decreased by one):
.nf
.RS
# i2ctransfer 0 w17@0x50 0x42 0xff-
.RE
.fi

.SH WARNING
.B i2ctransfer
can be extremely dangerous if used improperly.
It can confuse your I2C bus, cause data loss, or have more serious side effects.
Writing to a serial EEPROM on a memory DIMM (chip addresses between 0x50 and 0x57) may DESTROY your memory, leaving your system unbootable!
Be extremely careful using this program.

.SH BUGS
To report bugs or send fixes, please write to the Linux I2C mailing list
<linux-i2c@vger.kernel.org> with Cc to the current maintainer:
Jean Delvare <jdelvare@suse.de>.

.SH AUTHORS
Wolfram Sang, based on
.B i2cget
by Jean Delvare

This manual page was originally written by Wolfram Sang based on the manual
for
.B i2cset
by David Z Maze <dmaze@debian.org>.

.SH SEE ALSO
.BR i2cdetect (8), i2cdump (8), i2cget (8), i2cset (8)


CHANGES
i2c-tools CHANGES
-----------------

4.4 (2024-10-10)
  tools: Use getopt
         Implement and document option -h
  eeprog: Use force option when data comes from a pipe
  i2cdetect: Display more functionality bits with option -F
  i2cdump: Remove support for SMBus block mode
  i2cget: Document SMBus block mode
  i2ctransfer: Sort command line options and add to help text
               Add an option to print binary data
               Drop redundant variable arg_idx
  py-smbus: Install in the defined prefix
            Use setuptools instead of distutils

4.3 (2021-07-22)
  decode-dimms: Attempt to decode LPDDR3 modules
  eeprom, eepromer: Removed the tools in favor of eeprog
  i2cdetect: Sort the bus list by number
  i2cdump: Add range support to I2C block mode
           Deprecate SMBus block mode
  i2cget: Add support for I2C block read
          Add support for SMBus block read
  i2ctransfer: Reverted check for returned length from driver

4.2 (2020-09-22)
  manpages: Add BUGS section to let people know how to contact us
  Makefile: Allow to preset all CFLAGS and LDFLAGS variables
  tools: Consistently use snprintf instead of sprintf
         Restrict addresses 0x03-0x07, too (defined by I2C standard)
  decode-dimms: Print SPD revision for DDR3 too
                Print primary bus width for DDR3 and DDR4
                List ee1004 as a candidate driver
                Display MAC for DDR3
                Add MAC abbreviation for DDR4
                Round DDR4 speed properly
                Detect and report truncated input files
                Print kernel driver used
                Print DDR memory speed in MT/s
                Add DDR5 memory types
                Decode manufacturing data for LPDDR3
                Fix the version string
                Point the user to the right drivers
                Update the list of vendors to Jedec JEP106BB
  decode-vaio: Add support for the at24 driver
               Scan more i2c buses
  i2cset: Fix short writes with mask
  i2ctransfer: Mention '-a' everywhere in the manpage
               Support messages using I2C_M_RECV_LEN
               Add check for returned length from driver
  i2c-stub-from-dump: Read dumps from hexdump -C
  library: Add a manual page to document the API

4.1 (2018-11-30)
  Makefile: Make STRIP, DESTDIR and PREFIX overridable
  tools: Fix potential buffer overflows in i2cbusses
         Fix build race
         Allow usage of reserved addresses with the '-a' flag
  decode-dimms: Add preliminary DDR4 support
                Decode size and timings of DDR4
                Decode misc parameters of DDR4
                Decode physical characteristics of DDR4
                Documentation update for DDR4
                Verify the CRC of DDR4 data block 1
                Update manufacturer IDs (JEP106AX)
  eeprog: Fix ambiguous parentheses
          Fix build race
  i2ctransfer: Rename option '-f' to '-a' for consistency
  i2c-dev.h: Delete
  library: Fix build race
           Allow disabling the dynamic flavor
           Mention the correct license in source files
  py-smbus: Fix i2c_smbus_* error propagation

4.0 (2017-10-30)
  tools: Fix build with recent compilers (gcc 4.6+)
         Add examples to the manual pages
  README: Clarify licenses
          Mention the current maintainer
  decode-dimms: Decode module configuration type of DDR2 SDRAM
                Decode bus width extension of DDR3 SDRAM
                Don't choke when no EEPROM is found
                Don't make columns larger than they need to be
                Make side-by-side output more robust
                Print module organization of DDR SDRAM
                Merge cells by default in side-by-side output
                Print extra timing values of DDR SDRAM
                Print DDR and DDR2 core timings for all supported CAS values
                Print DDR2 equivalent speed of tCK max
                Don't print undefined DDR2 SDRAM timings
		Print SDR, DDR, DDR2, DDR3 core timings for all standard speeds
                Update manufacturer IDs
                Make DDR3 manufacturer count parity error non-fatal
                Strip former manufacturer name in side-by-side output mode
                Remove duplicate "ns" in SDR timings
                Add section headers for SDR modules
                Fix decoding of SDR SPD revision
                Prevent hang on reserved DDR3 module type
                Decode more DDR3 module types
                Fix DDR3 tRAS decoding
                Fix DDR3 core timings rounding
                Round down PC3 numbers to comply with Jedec
                Don't print the DDR3 time bases
                Decode the FTB fields of DDR3 tCk, tAA, tRCD, tRP and tRC
                Fix speed and PC3 number of high-speed DDR3 modules
                Decode DDR3 reference card revision
                Print width of all known DDR3 module types
                Print physical characteristics for all DDR3 module types
                Don't print raw SSTE32882 register values
                Add support for Load Reduced DIMM (LRDIMM) DDR3 modules
                Fully decode the DDR3 SDRAM Device Type field
                Fix DDR3 extended temp range refresh rate decoding
                Encode "degrees" to HTML degree symbol
                Generate XHTML 1.1 compliant markup
                Add a manual page
                Correctly check for out-of-bounds vendor ID
                Update manufacturer IDs (JEP106AQ)
  decode-vaio: Add a manual page
  eeprog: Add a manual page
          Moved to a separate subdirectory
          Increase delay after writes
  eeprom: Add a manual page
          Marked as deprecated
  eepromer: Add a manual page
            Marked as deprecated
  i2cdetect: Do a best effort detection if functionality is missing
             Clarify the SMBus commands used for probing by default
  i2ctransfer: New tool to send user-defined I2C messages in one transfer

tools/i2ctransfer.c
/*
    i2ctransfer.c - A user-space program to send concatenated i2c messages
    Copyright (C) 2015-17 Wolfram Sang <wsa@sang-engineering.com>
    Copyright (C) 2015-17 Renesas Electronics Corporation

    Based on i2cget.c:
    Copyright (C) 2005-2012  Jean Delvare <jdelvare@suse.de>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
*/

#include <sys/ioctl.h>
#include <errno.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <linux/i2c.h>
#include <linux/i2c-dev.h>
#include "i2cbusses.h"
#include "util.h"
#include "../version.h"

enum parse_state {
	PARSE_GET_DESC,
	PARSE_GET_DATA,
};

enum supported_flags_retval {
	FLAGS_GOT_RW,
	FLAGS_GOT_OPTIONAL,
	FLAGS_UNEXPECTED,
	FLAGS_UNKNOWN,
	FLAGS_UNSUPPORTED,
};

#define PRINT_STDERR	(1 << 0)
#define PRINT_READ_BUF	(1 << 1)
#define PRINT_WRITE_BUF	(1 << 2)
#define PRINT_HEADER	(1 << 3)
#define PRINT_BINARY	(1 << 4)

static void help(void)
{
	fprintf(stderr,
		"Usage: i2ctransfer [OPTIONS] I2CBUS DESC [DATA] [DESC [DATA]]...\n"
		"  OPTIONS: -a allow even reserved addresses\n"
		"           -b print read data as binary, disables -v\n"
		"           -f force access even if address is marked used\n"
		"           -h this help text\n"
		"           -v verbose mode\n"
		"           -V version info\n"
		"           -y yes to all confirmations\n"
		"  I2CBUS is an integer or an I2C bus name\n"
		"  DESC describes the transfer in the form: [inpst]{r|w}LENGTH[@address]\n"
		"    1) optional message modifier flags, if supported\n"
		"       i: ignore NACK from client\n"
		"       n: no master ACK/NACK bit in a read message\n"
		"       p: emit a STOP after the message\n"
		"       s: skip repeated start\n"
		"       t: toggle read/write bit\n"
		"    2) mandatory read/write flag\n"
		"    3) LENGTH (range 0-65535, or '?')\n"
		"    4) I2C address (use last one if omitted)\n"
		"  DATA are LENGTH bytes for a write message. They can be shortened by a suffix:\n"
		"    = (keep value constant until LENGTH)\n"
		"    + (increase value by 1 until LENGTH)\n"
		"    - (decrease value by 1 until LENGTH)\n"
		"    p (use pseudo random generator until LENGTH with value as seed)\n\n"
		"Example (bus 0, read 8 byte at offset 0x64 from EEPROM at 0x50):\n"
		"  # i2ctransfer 0 w1@0x50 0x64 r8\n"
		"Example (same EEPROM, at offset 0x42 write 0xff 0xfe ... 0xf0):\n"
		"  # i2ctransfer 0 w17@0x50 0x42 0xff-\n");
}

static int get_funcs(int file, unsigned long *funcs)
{
	/* check adapter functionality */
	if (ioctl(file, I2C_FUNCS, funcs) < 0) {
		fprintf(stderr, "Error: Could not get the adapter "
			"functionality matrix: %s\n", strerror(errno));
		return -1;
	}

	return 0;
}

static int check_funcs(unsigned long funcs)
{
	if (!(funcs & I2C_FUNC_I2C)) {
		fprintf(stderr, MISSING_FUNC_FMT, "I2C transfers");
		return -1;
	}

	return 0;
}

static void print_msgs(struct i2c_msg *msgs, __u32 nmsgs, unsigned int flags)
{
	FILE *output = flags & PRINT_STDERR ? stderr : stdout;
	unsigned int i;
	__u16 j;

	for (i = 0; i < nmsgs; i++) {
		int read = msgs[i].flags & I2C_M_RD;
		int recv_len = msgs[i].flags & I2C_M_RECV_LEN;
		int print_buf = (read && (flags & PRINT_READ_BUF)) ||
				(!read && (flags & PRINT_WRITE_BUF));
		__u16 len = recv_len ? msgs[i].buf[0] + 1 : msgs[i].len;

		if (flags & PRINT_HEADER) {
			fprintf(output, "msg %u: addr 0x%02x, %s, len ",
				i, msgs[i].addr, read ? "read" : "write");
			if (!recv_len || flags & PRINT_READ_BUF)
				fprintf(output, "%u", len);
			else
				fprintf(output, "TBD");
		}

		if (len && print_buf) {
			if (flags & PRINT_BINARY) {
				fwrite(msgs[i].buf, 1, len, output);
			} else {
				if (flags & PRINT_HEADER)
					fprintf(output, ", buf ");
				for (j = 0; j < len - 1; j++)
					fprintf(output, "0x%02x ", msgs[i].buf[j]);
				/* Print final byte with newline */
				fprintf(output, "0x%02x\n", msgs[i].buf[j]);
			}
		} else if (flags & PRINT_HEADER) {
			fprintf(output, "\n");
		}
	}
}

static int confirm(const char *filename, struct i2c_msg *msgs, __u32 nmsgs)
{
	fprintf(stderr, "WARNING! This program can confuse your I2C bus, cause data loss and worse!\n");
	fprintf(stderr, "I will send the following messages to device file %s:\n", filename);
	print_msgs(msgs, nmsgs, PRINT_STDERR | PRINT_HEADER | PRINT_WRITE_BUF);

	fprintf(stderr, "Continue? [y/N] ");
	fflush(stderr);
	if (!user_ack(0)) {
		fprintf(stderr, "Aborting on user request.\n");
		return 0;
	}

	return 1;
}

/*
 * Parse one message flag and check if the adapter supports it.
 * Return:
 * FLAGS_GOT_RW if we have the direction of the message (read or write)
 * FLAGS_GOT_OPTIONAL if the argument is a valid and supported optional flag
 * FLAGS_UNEXPECTED if the argument is the beginning of the size or address
 * FLAGS_UNKNOWN for unknown flag
 * FLAGS_UNSUPPORTED for flag unsupported by the adapter
 */
static enum supported_flags_retval
	add_flag_if_supported(__u16 *flags, unsigned long funcs, char arg)
{
	int mangling = 0, nostart = 0;

	/* Get the message flag from the argument */
	switch (arg) {
	/* mandatory direction flag: ends the parsing of flags */
	case 'r':
		*flags |= I2C_M_RD;
		return FLAGS_GOT_RW;
	case 'w':
		return FLAGS_GOT_RW;

	/* optional flags */
#ifdef I2C_M_IGNORE_NAK
	case 'i':
		*flags |= I2C_M_IGNORE_NAK;
		mangling = 1;
		break;
#endif
#ifdef I2C_M_NO_RD_ACK
	case 'n':
		*flags |= I2C_M_NO_RD_ACK;
		mangling = 1;
		break;
#endif
#ifdef I2C_M_STOP
	case 'p':
		*flags |= I2C_M_STOP;
		mangling = 1;
		break;
#endif
#ifdef I2C_M_NOSTART
	case 's':
		*flags |= I2C_M_NOSTART;
		nostart = 1;
		break;
#endif
#ifdef I2C_M_REV_DIR_ADDR
	case 't':
		*flags |= I2C_M_REV_DIR_ADDR;
		mangling = 1;
		break;
#endif

	/* unexpected next part of the message: length or address */
	case '0':
	case '1':
	case '2':
	case '3':
	case '4':
	case '5':
	case '6':
	case '7':
	case '8':
	case '9':
	case '?':
	case '@':
		return FLAGS_UNEXPECTED;

	default:
		fprintf(stderr, "Error: Unknown flag '%c'\n", arg);
		return FLAGS_UNKNOWN;
	}

	/* Check that the adapter supports the requested flags */
#ifdef I2C_FUNC_PROTOCOL_MANGLING
	if (funcs & I2C_FUNC_PROTOCOL_MANGLING)
		mangling = 0;
#endif
	if (mangling) {
		fprintf(stderr, MISSING_FUNC_FMT, "protocol mangling");
		return FLAGS_UNSUPPORTED;
	}

#ifdef I2C_FUNC_NOSTART
	if (funcs & I2C_FUNC_NOSTART)
		nostart = 0;
#endif
	if (nostart) {
		fprintf(stderr, MISSING_FUNC_FMT, "repeated start skipping");
		return FLAGS_UNSUPPORTED;
	}

	return FLAGS_GOT_OPTIONAL;
}

int main(int argc, char *argv[])
{
	char filename[20];
	int i2cbus, address = -1, file, opt, nmsgs = 0, nmsgs_sent, i;
	int force = 0, yes = 0, version = 0, verbose = 0, all_addrs = 0, binary = 0;
	struct i2c_msg msgs[I2C_RDRW_IOCTL_MAX_MSGS];
	enum parse_state state = PARSE_GET_DESC;
	unsigned int buf_idx = 0;
	unsigned long funcs;

	memset(msgs, 0, sizeof(msgs));

	/* handle (optional) flags first */
	while ((opt = getopt(argc, argv, "abfhvVy")) != -1) {
		switch (opt) {
		case 'a': all_addrs = 1; break;
		case 'b': binary = 1; break;
		case 'f': force = 1; break;
		case 'v': verbose = 1; break;
		case 'V': version = 1; break;
		case 'y': yes = 1; break;
		case '?':
		case 'h':
			help();
			exit(opt == '?');
		}
	}

	if (version) {
		fprintf(stderr, "i2ctransfer version %s\n", VERSION);
		exit(0);
	}

	if (optind == argc) {
		help();
		exit(1);
	}

	i2cbus = lookup_i2c_bus(argv[optind++]);
	if (i2cbus < 0)
		exit(1);

	file = open_i2c_dev(i2cbus, filename, sizeof(filename), 0);
	if (file < 0 || get_funcs(file, &funcs) || check_funcs(funcs))
		exit(1);

	while (optind < argc) {
		char *arg_ptr = argv[optind];
		unsigned long len, raw_data;
		__u16 flags;
		__u8 data, *buf;
		char *end;
		int ret;

		if (nmsgs == I2C_RDRW_IOCTL_MAX_MSGS) {
			fprintf(stderr, "Error: Too many messages (max: %d)\n",
				I2C_RDRW_IOCTL_MAX_MSGS);
			goto err_out;
		}

		switch (state) {
		case PARSE_GET_DESC:
			flags = 0;

			for (ret = FLAGS_GOT_OPTIONAL; *arg_ptr && ret == FLAGS_GOT_OPTIONAL; arg_ptr++)
				ret = add_flag_if_supported(&flags, funcs, *arg_ptr);
			if (ret == FLAGS_UNKNOWN || ret == FLAGS_UNSUPPORTED)
				goto err_out_with_arg;
			if (ret != FLAGS_GOT_RW) {
				fprintf(stderr, "Error: Missing direction flag 'r' or 'w'\n");
				goto err_out_with_arg;
			}

			if (*arg_ptr == '?') {
				if (!(flags & I2C_M_RD)) {
					fprintf(stderr, "Error: variable length not allowed with write\n");
					goto err_out_with_arg;
				}
				len = 256; /* SMBUS3_MAX_BLOCK_LEN + 1 */
				flags |= I2C_M_RECV_LEN;
				arg_ptr++;
			} else {
				len = strtoul(arg_ptr, &end, 0);
				if (len > 0xffff || arg_ptr == end) {
					fprintf(stderr, "Error: Length invalid\n");
					goto err_out_with_arg;
				}
				arg_ptr = end;
			}

			if (*arg_ptr) {
				if (*arg_ptr++ != '@') {
					fprintf(stderr, "Error: Unknown separator after length\n");
					goto err_out_with_arg;
				}

				/* We skip 10-bit support for now. If we want it,
				 * it should be marked with a 't' flag before
				 * the address here.
				 */

				address = parse_i2c_address(arg_ptr, all_addrs);
				if (address < 0)
					goto err_out_with_arg;

				/* Ensure address is not busy */
				if (!force && set_slave_addr(file, address, 0))
					goto err_out_with_arg;
			} else {
				/* Reuse last address if possible */
				if (address < 0) {
					fprintf(stderr, "Error: No address given\n");
					goto err_out_with_arg;
				}
			}

			msgs[nmsgs].addr = address;
			msgs[nmsgs].flags = flags;
			msgs[nmsgs].len = len;

			if (len) {
				buf = malloc(len);
				if (!buf) {
					fprintf(stderr, "Error: No memory for buffer\n");
					goto err_out_with_arg;
				}
				memset(buf, 0, len);
				msgs[nmsgs].buf = buf;
				if (flags & I2C_M_RECV_LEN)
					buf[0] = 1; /* number of extra bytes */
			}

			if (flags & I2C_M_RD || len == 0) {
				nmsgs++;
			} else {
				buf_idx = 0;
				state = PARSE_GET_DATA;
			}

			break;

		case PARSE_GET_DATA:
			raw_data = strtoul(arg_ptr, &end, 0);
			if (raw_data > 0xff || arg_ptr == end) {
				fprintf(stderr, "Error: Invalid data byte\n");
				goto err_out_with_arg;
			}
			data = raw_data;
			len = msgs[nmsgs].len;

			while (buf_idx < len) {
				msgs[nmsgs].buf[buf_idx++] = data;

				if (!*end)
					break;

				switch (*end) {
				/* Pseudo randomness (8 bit AXR with a=13 and b=27) */
				case 'p':
					data = (data ^ 27) + 13;
					data = (data << 1) | (data >> 7);
					break;
				case '+': data++; break;
				case '-': data--; break;
				case '=': break;
				default:
					fprintf(stderr, "Error: Invalid data byte suffix\n");
					goto err_out_with_arg;
				}
			}

			if (buf_idx == len) {
				nmsgs++;
				state = PARSE_GET_DESC;
			}

			break;

		default:
			/* Should never happen */
			fprintf(stderr, "Internal Error: Unknown state in state machine!\n");
			goto err_out;
		}

		optind++;
	}

	if (state != PARSE_GET_DESC || nmsgs == 0) {
		fprintf(stderr, "Error: Incomplete message\n");
		goto err_out;
	}

	if (yes || confirm(filename, msgs, nmsgs)) {
		struct i2c_rdwr_ioctl_data rdwr;
		unsigned int print_flags = PRINT_READ_BUF;

		memset(&rdwr, 0, sizeof(rdwr));
		rdwr.msgs = msgs;
		rdwr.nmsgs = nmsgs;
		nmsgs_sent = ioctl(file, I2C_RDWR, &rdwr);
		if (nmsgs_sent < 0) {
			fprintf(stderr, "Error: Sending messages failed: %s\n", strerror(errno));
			goto err_out;
		} else if (nmsgs_sent < nmsgs) {
			fprintf(stderr, "Warning: only %d/%d messages were sent\n", nmsgs_sent, nmsgs);
		}

		if (binary)
			print_flags |= PRINT_BINARY;
		else if (verbose)
			print_flags |= PRINT_HEADER | PRINT_WRITE_BUF;

		print_msgs(msgs, nmsgs_sent, print_flags);
	}

	close(file);

	for (i = 0; i < nmsgs; i++)
		free(msgs[i].buf);

	exit(0);

 err_out_with_arg:
	fprintf(stderr, "Error: faulty argument is '%s'\n", argv[optind]);
 err_out:
	close(file);

	/*
	 * If we were parsing data, the buffer for the last message was
	 * already allocated and nmsgs still points to it.
	 */
	if (state == PARSE_GET_DATA)
		free(msgs[nmsgs].buf);
	for (i = 0; i < nmsgs; i++)
		free(msgs[i].buf);

	exit(1);
}
